README.md
# Cloudflare Access for Flask

Zero Trust Access with Cloudflare Access for Flask applications.


## Links

- Documentation: <https://flask-cfaccess.readthedocs.io/>
- Changes: <https://flask-cfaccess.readthedocs.io/changes/>
- Releases: <https://pypi.org/project/flask-cfaccess/>
- Source Code: <https://github.com/COUR4G3/flask-cfaccess>
- Issue Tracker: <https://github.com/COUR4G3/flask-cfaccess/issues>


CHANGES.md
# Changes

## Version 0.1.1

Release 2023-04-06

* Fixed missing ``jwt`` and ``requests`` dependencies.


## Version 0.1.0

Released 2023-04-06

* Initial release.


tests/conftest.py
import os
import secrets
import time

from pathlib import Path

import jwt
import pytest
import responses

from flask import Flask
from flask_cfaccess import CfAccess

try:
    import flask_login
except ImportError:
    flask_login = None


algo = jwt.algorithms.RSAAlgorithm(jwt.algorithms.RSAAlgorithm.SHA256)

key_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "keys")

with open(os.path.join(key_dir, "rsa_private_key.json")) as private:
    PRIVATE_KEY = algo.from_jwk(private.read())

with open(os.path.join(key_dir, "rsa_public_key.json")) as public:
    PUBLIC_KEY = algo.from_jwk(public.read())


@pytest.fixture
def app(mocker, tmp_path: Path, token_payload):
    app = Flask("flask_cfaccess", instance_path=str(tmp_path / "instance"))

    app.config[
        "CLOUDFLARE_ACCESS_AUD"
    ] = "32eafc7626e974616deaf0dc3ce63d7bcbed58a2731e84d06bc3cdf1b53c4228"
    app.config["CLOUDFLARE_ACCESS_TEAM"] = "yourteam"

    app.secret_key = secrets.token_hex(16)
    app.testing = True

    CfAccess(app)

    jwks_client = app.extensions["cfaccess"][1]

    def get_public_key(token):
        return PUBLIC_KEY

    mocker.patch.object(
        jwks_client, "get_signing_key_from_jwt", get_public_key
    )

    responses.get(
        url="https://yourteam.cloudflareaccess.com/cdn-cgi/access/get-identity",
        json=token_payload,
    )

    return app


@pytest.fixture()
def client(app):
    return app.test_client()


@pytest.fixture
def app_context(app: Flask):
    with app.app_context() as ctx:
        yield ctx


@pytest.fixture(scope="session")
def token(token_payload):
    return jwt.encode(token_payload, PRIVATE_KEY, "RS256")


@pytest.fixture(scope="session")
def token_payload():
    now = time.time()

    return {
        "aud": [
            "32eafc7626e974616deaf0dc3ce63d7bcbed58a2731e84d06bc3cdf1b53c4228"
        ],
        "email": "user@example.com",
        "exp": now + 300,
        "iat": now - 300,
        "nbf": now - 300,
        "iss": "https://yourteam.cloudflareaccess.com",
        "type": "app",
        "identity_nonce": "6ei69kawdKzMIAPF",
        "sub": "7335d417-61da-459d-899c-0a01c76a2f94",
        "country": "US",
    }


if flask_login:

    @pytest.fixture
    def login_manager(app: Flask):
        return flask_login.LoginManager(app)


def pytest_addoption(parser):
    parser.addoption(
        "--integration",
        action="store_true",
        default=False,
        help="run integration tests",
    )


def pytest_collection_modifyitems(config, items):
    if config.getoption("--integration"):
        # --integration given in cli: do not skip integration tests
        return
    skip_integration = pytest.mark.skip(
        reason="need --integration option to run",
    )
    for item in items:
        if "integration" in item.keywords:
            item.add_marker(skip_integration)


tests/test_cfaccess.py
from dataclasses import dataclass

import jwt
import pytest
import responses

from flask_cfaccess import (
    CfAccess,
    cfaccess,
    login_required,
    validate_identity,
)

try:
    import flask_login
except ImportError:
    flask_login = None


FAKE_PRIVATE_KEY = """-----BEGIN PRIVATE KEY-----
MIIEvwIBADANBgkqhkiG9w0BAQEFAASCBKkwggSlAgEAAoIBAQCotVge0KSFI1+b
eq8yKMjT19bHIdITneE9nWcZyNsrJ+Z1GU/+WOo0KvbjfjRpcXTkE9uwHLHmupVT
x+QlWk5fR14u79hCn20m4jnOHMeY/5xu8XHrIqcd/GiD9Wqd4P1hqqCeCKu3V1ty
cMyfWuPvpP8fc0N/5qIxZuFHs2pJNRpIucWFZ0FtWAYUTkKwt//k5bFPVpXFh3rE
NqqYiem7qSd4y7ROuVzjOKQKxvUgkhCeQteggaMBjkv3WtIzxYIEj8Eu3pa76rsW
F8mqSKM28xNryvdO9mUiaLgT7OYlhJgu8n5lIwwfqZGbyxdOYBy/p1bikJPcBiJH
Jx5sdOHvAgMBAAECggEATXcrVcbu7DVQZreAL5N7l45RkBEvAZPqg1y9pSDNxjSl
hNpRGVShgb/CqrXeAha3Gi0mIGkkpPHCtHwh/abIOsat7/ZwgjLrL0IMUeB0JAQm
w+h4qwO34bRvFzx+rqzLG+f5pWcM1fw2QMebpYZaCEVmYf4OnfF8Qxve0IoJAUCx
1hbIG8TN5+i5aQsQ8T3eaq/cOuDTZKelAPIsIEQf/H6acUGEDC+XumP5/UywoD2A
5OKXMERDr1IqiE622vy8TblFhct5TTC34zZyp46lXN3ujeTCE8d5tWjxUqr58Sxg
n94SWgoykH2PfyOmuMZulj+DJcVLXMIcAWyo/JJiQQKBgQC+lR/OgYQE6CD/wzDV
QF7mauBXXH/LYWlNL1PMham4UC02hd+ppbwpVt2/+Xbn5ZCGVuNoRzntuxpXvREB
1TQofuM6s0Ocxa9gc3E5k/OT3YMENQTGXt9v+zzi8jBjmHryI480r8kZVfxintkF
A5QRmd4WXLs+MM3QUKA+/igLrwKBgQDinhgu4GwZ+KE0D8A2BnUdxslhmURrWHvt
4kYZFFus/jbB+meGHzpkyIP6GlmX0OiYn2rMwPK0npM12glwdn5ECsVrVPQQ9IR6
/iIXY28QBovYI/VD9Zw/PPPOmgUfFJH0lHlQsRjulWzAODngeJBfJBV46fmmfGFQ
S5nI6jLdwQKBgQCiu4LlOgIYjn1e5rST4tsltKzK69jTOOJJXdgYEynQj+WLYNSS
b0ozolDG06SMbqDxlM+pAuib7d+Ye4BCKGLZET/Fp7xb6RDPMsPDfmnGIKb2uK7T
E7+GRui+gD6BTHnuuB5XxqHYWGuhajG07lFIljHktlnLyLB05+Ns5mgbUQKBgQCl
6/ro14hOu6vL4nxpEfFOC+4Gd6GQtwiF89Y/Jpi44KhW/nAt00992xJGXGScs2iZ
1p7rq5oIhNJj63Bd12Q7jWdeXqUUhPM62jc7yKHYqfJaHnnFeF8eLhKUglCdoQH8
gXkRi5kllmI5mvFFWQDbqB/Ql/8h1o+pqrC+WjtEAQKBgQCRb3nuvlXg/dQrw1tt
vEAbOaQGMJRiRinadSHdSCpp+BNjbycKfeA7mwIQLZOeMjYw1175eKZsjQWuRmCz
kF0iWTxbimpV44KVaWenxchMTKGC9MCywS7oXwDnS8xjOBvmrhkWQh3bUaRzcG+t
C2/akUdFVqOhJ/Ux3WZHOk+QkA==
-----END PRIVATE KEY-----"""


def test_authenticate(app, token, token_payload):
    with app.test_request_context(headers={"Cf-Access-Jwt-Assertion": token}):
        assert cfaccess.authenticate()


def test_authenticate_bad_aud(app):
    app.config["CLOUDFLARE_ACCESS_AUD"] = "wrong"

    with app.test_request_context():
        assert not cfaccess.authenticate()


def test_authenticate_bad_token(app, token_payload):
    token = jwt.encode(token_payload, FAKE_PRIVATE_KEY, "RS256")

    with app.test_request_context(headers={"Cf-Access-Jwt-Assertion": token}):
        assert not cfaccess.authenticate()


def test_authenticate_missing_token(app, token, token_payload):
    with app.test_request_context():
        assert not cfaccess.authenticate()


def test_init_app(app):
    cfaccess = CfAccess()
    cfaccess.init_app(app)

    app.extensions["cfaccess"] is cfaccess


@responses.activate
def test_get_identity(app, mocker, token, token_payload):
    with app.test_request_context():
        assert not cfaccess.get_identity()
    with app.test_request_context(headers={"Cf-Access-Jwt-Assertion": token}):
        identity = cfaccess.get_identity()
        assert identity == token_payload


def test_login_required(app, client, token, token_payload):
    @app.route("/")
    @login_required
    def protected():
        return "Protected"

    resp = client.get("/")
    assert resp.status_code == 401

    headers = {
        "Cookie": f"CF_Authorization={token}",
        "Cf-Access-Jwt-Assertion": token,
    }

    resp = client.get("/", headers=headers)
    assert resp.status_code == 200


@responses.activate
def test_validate_identity(app, client, token):
    def identify(user_info):
        return user_info["email"] == "user@example.com"

    @app.route("/")
    @validate_identity(identify)
    def protected():
        return "Protected"

    headers = {
        "Cookie": f"CF_Authorization={token}",
        "Cf-Access-Jwt-Assertion": token,
    }

    resp = client.get("/", headers=headers)
    assert resp.status_code == 200


@responses.activate
def test_validate_identity_wrong_user(app, client, token):
    def identify(user_info):
        return user_info["email"] == "admin@example.com"

    @app.route("/")
    @validate_identity(identify)
    def protected():
        return "Protected"

    headers = {
        "Cookie": f"CF_Authorization={token}",
        "Cf-Access-Jwt-Assertion": token,
    }

    resp = client.get("/", headers=headers)
    assert resp.status_code == 403


def test_token_cookie_only(app, token):
    with app.test_request_context(headers={"Cf-Access-Jwt-Assertion": token}):
        assert cfaccess.authenticate()


def test_token_header_only(app, token):
    with app.test_request_context(
        headers={"Cookie": f"CF_Authorization={token}"}
    ):
        assert cfaccess.authenticate()


if flask_login:

    @dataclass
    class User(flask_login.UserMixin):
        email: str

        def get_id(self):
            return self.email

    @pytest.fixture
    def user():
        return User(email="user@example.com")

    @pytest.fixture
    def users(user):
        return {user.email: user}


def _test_flask_login(app, login_manager, users):
    @login_manager.request_loader
    def request_loader(request):
        identity = cfaccess.get_identity()
        if not identity:
            return
        return users.get(identity["email"])

    @app.route("/")
    @flask_login.login_required
    def protected():
        return "Protected"


@pytest.mark.skipif(not flask_login, reason="flask-login not installed")
@responses.activate
def test_flask_login(app, client, login_manager, token, user, users):
    _test_flask_login(app, login_manager, users)

    headers = {
        "Cookie": f"CF_Authorization={token}",
        "Cf-Access-Jwt-Assertion": token,
    }

    resp = client.get("/", headers=headers)
    assert resp.status_code == 200


@pytest.mark.skipif(not flask_login, reason="flask-login not installed")
@responses.activate
def test_flask_login_missing_token(
    app, client, login_manager, token, user, users
):
    _test_flask_login(app, login_manager, users)

    resp = client.get("/")
    assert resp.status_code == 401


@pytest.mark.skipif(not flask_login, reason="flask-login not installed")
@responses.activate
def test_flask_login_missing_user(
    app, client, login_manager, token, user, users
):
    _test_flask_login(app, login_manager, users)

    users.pop(user.email)

    headers = {
        "Cookie": f"CF_Authorization={token}",
        "Cf-Access-Jwt-Assertion": token,
    }

    resp = client.get("/", headers=headers)
    assert resp.status_code == 401


flask_cfaccess.py
"""Zero Trust Access with Cloudflare Access for Flask applications."""
import typing as t

from functools import wraps
from importlib.metadata import PackageNotFoundError, version

import jwt
import requests

from flask import Flask, abort, current_app, g, request
from jwt.jwks_client import PyJWKClient
from werkzeug.local import LocalProxy

try:
    from _version import version as __version__
except ImportError:  # pragma: nocover
    try:
        __version__ = version("flask_cfaccess")
    except PackageNotFoundError:
        __version__ = "0.1-dev0"


class CfAccess:
    """Zero Trust Access with Cloudflare Access for Flask applications.

    :param app: The Flask application instance. The application instance is
                optional and can be passed by calling
                ``cfaccess.init_app(app)`` in your application factory\
                function.

    """

    def __init__(self, app: t.Optional[Flask] = None):
        self.default_app = app
        if app:
            self.init_app(app)

    def init_app(self, app: Flask):
        """Initialize the application.

        :param app: The Flask application.instance.

        """
        try:
            team = app.config["CLOUDFLARE_ACCESS_AUD"]
            team = app.config["CLOUDFLARE_ACCESS_TEAM"]
        except KeyError:
            raise RuntimeError(
                "CLOUDFLARE_ACCESS_AUD and CLOUDFLARE_ACCESS_TEAM is required "
                "for flask-cfaccess"
            )

        client = PyJWKClient(
            f"https://{team}.cloudflareaccess.com/cdn-cgi/access/certs"
        )

        if not hasattr(app, "extensions"):
            app.extensions = {}

        app.extensions["cfaccess"] = (self, client)

    def _get_app(self):
        try:
            return current_app._get_current_object()
        except RuntimeError:
            if self.default_app:
                return self.default_app

            raise

    def authenticate(self, token=None):
        """Validate that a request is authenticated."""
        if not token:
            token = request.headers.get("Cf-Access-Jwt-Assertion")
        if not token:
            token = request.cookies.get("CF_Authorization")
        if not token:
            return False

        app = self._get_app()

        try:
            cfaccess, jwks_client = app.extensions["cfaccess"]
        except (AttributeError, IndexError, KeyError, TypeError):
            raise RuntimeError(
                "flask-cfaccess not configured for this application"
            )

        aud = app.config["CLOUDFLARE_ACCESS_AUD"]
        key = jwks_client.get_signing_key_from_jwt(token)

        try:
            payload = jwt.decode(
                token, key, audience=aud, algorithms=["RS256"]
            )
        except jwt.DecodeError:
            return False

        g._cfaccess_token = token
        g._cfaccess_payload = payload

        return True

    def get_identity(self, token=None):
        """Get the user's identity."""
        if not token:
            token = getattr(g, "_cfaccess_token", None)

            if not token:
                self.authenticate()

            token = getattr(g, "_cfaccess_token", None)
            if not token:
                return

        team = self._get_app().config["CLOUDFLARE_ACCESS_TEAM"]
        url = (
            f"https://{team}.cloudflareaccess.com/cdn-cgi/access/get-identity"
        )

        res = requests.get(url, cookies={"CF_Authorization": token})
        res.raise_for_status()

        return res.json()

    def login_required(self, f):
        """Decorate a view to require a valid identity."""
        return login_required(f)

    def validate_identity(self, identify):
        """Decorate a view to validate the user's identity."""
        return validate_identity(identify)


def get_cfaccess():
    try:
        return current_app.extensions["cfaccess"][0]
    except (AttributeError, IndexError, KeyError, TypeError):
        raise RuntimeError(
            "flask-cfaccess not configured for this application"
        )


cfaccess = LocalProxy(get_cfaccess)


def login_required(f):
    """Decorate a view to require that the request is authenticated."""

    @wraps(f)
    def wrapper(*args, **kwargs):

        if not cfaccess.authenticate():
            abort(401)

        return f(*args, **kwargs)

    return wrapper


def get_identity():
    """Get the user's identity."""

    return cfaccess.get_identity()


def validate_identity(identify):
    """Decorate a view to validate the user's identity."""

    def decorator(f):
        @wraps(f)
        def wrapper(*args, **kwargs):
            if not cfaccess.authenticate():
                abort(401)

            identity = cfaccess.get_identity()
            if not identify(identity):
                abort(403)

            return f(*args, **kwargs)

        return wrapper

    return decorator
